boltstub/wiring.py

```diff
@@ -129,6 +135,101 @@ def __str__(self):
 
 
+class RegularSocket:
+    """A socket with an already receive cached value."""
+
+    def __init__(self, _socket, _cache) -> None:
+        self._socket = _socket
+        self._cache = _cache
+
+    def __getattr__(self, item):
+        return getattr(self._socket, item)
+
+    def recv(self, __bufsize) -> bytes:
+        if (self._cache):
+            buff = self._cache
+            self._cache = None
+            return buff
+        return self._socket.recv(__bufsize)
+
+
+class WebSocket:
+    """Implementation of Websockets [rfc6455].
+
+    This implementation doesn't support extensions
+    """
+
+    def __init__(self, _socket) -> None:
+        self._socket = _socket
+
+    def __getattr__(self, item):
+        return getattr(self._socket, item)
+
+    def recv(self, __bufsize) -> bytes:
+        """Receives data from the socket.
+
+        The `__bufsize` parameter is ignored. This method returns the entire
+        frame payload and handles control frames doing the needed actions.
+        """
+        frame = self._socket.recv(2)
+        if len(frame) == 0:
+            return None
+
+        fin = frame[0] >> 7
+        # rsv1 = frame[0] & 0b0100_0000 == 0b0100_0000
+        # rsv2 = frame[0] & 0b0010_0000 == 0b0010_0000
+        # rsv3 = frame[0] & 0b0001_0000 == 0b0001_0000
+        opcode = frame[0] & 0b0000_1111
+
+        masked = frame[1] >> 7
+        payload_len = frame[1] & 0b0111_1111
+
+        if payload_len == 126:
+            payload_len, = struct.unpack(">H", self._socket.recv(2))
+        elif payload_len == 127:
+            payload_len, = struct.unpack(">Q", self._socket.recv(8))
+
+        if masked == 1:
+            mask = self._socket.recv(4)
+
+        masked_payload = self._socket.recv(payload_len)
+
+        payload = masked_payload \
+            if masked == 0 \
+            else bytearray([masked_payload[i] ^ mask[i % 4]
+                            for i in range(payload_len)])
+        if opcode & 0b0000_1000 == 0b0000_1000:
+            if opcode == 0x09:  # PING
+                self._socket.sendall(PONG)
+            return self.recv(__bufsize)
+        elif fin == 0:
+            return payload + self.recv(__bufsize)
+        return payload
+
+    def send(self, payload) -> int:
+        """Send the payload over the socket inside a Websocket frame."""
+        frame = [0b1000_0010]
+        payload_len = len(payload)
+        if (payload_len < 126):
+            frame += [payload_len]
+        elif payload_len < 0x10000:
+            frame += [126]
+            frame += bytearray(struct.pack(">H", payload_len))
+        else:
+            frame += [127]
+            frame += bytearray(struct.pack(">Q", payload_len))
+
+        frame_to_send = bytearray(frame) + bytearray(payload)
+
+        self._socket.sendall(frame_to_send)
+
+        return len(payload)
+
+    def sendall(self, payload):
+        """Send the payload over the socket inside a Websocket frame."""
+        self.send(payload)
```

Comment:
```suggestion
    sendall = send
```
for the lazy and idiomatic ones ;)